### ISSUE TITLE
reportMissingSuperCall and multiple inheritance

### ISSUE THREAD

**Issue by NeilGirdhar** [microsoft/pyright#2865]
The new reportMissingSuperCall feature is awesome!!  It doesn't catch all of the errors though:
```python
from typing import Any

class X:
    def __init__(self) -> None:
        self.x = 1

class Y(X):
    def __init__(self) -> None:   #  error (good)
        pass

class X0:
    pass

class Y0(X0):
    def __init__(self) -> None:  # no error!
        pass

class S(X0):
    def __init__(self, *args: Any, **kwargs: Any) -> None:
        super.__init__(*args, **kwargs)
        self.s = 1

class D(Y0, S):
    pass  # S.__init__ is not called!
```
There are two ways of implementing this feature.  Either
- the conservative approach whereby when D is defined, Y0 can no longer get away with not calling super, or
- the strict approach whereby all non-final classes must call super in case a common base like D is eventually defined.

Personally, I like the strict approach since it's more defensive.  A library may not realize that someone down the line uses multiple inheritance.  A concrete example is https://github.com/networkx/networkx/blob/main/networkx/classes/graph.py#L289.  `networkx.Graph` cannot be part of multiple inheritance because it doesn't call super.

**Comment by erictraut** [comment]
I don't generate an error when the only parent class in the MRO that provides an `__init__` method is `object` because its `__init__` method is known to have no side effects, and it's therefore safe not to call it. That explains why `Y0.__init__` doesn't emit an error.

The current check is applied to a present `__init__` method. Adding a check for a completely missing `__init__` method in the multiple inheritance case is an interesting idea, but that would involve very different logic.

So I don't consider either of these bugs. The first one is as designed. The second I consider another feature request. It's a bit more esoteric than the current check, but it's still a reasonable request.

**Comment by NeilGirdhar** [comment]
> So I don't consider either of these bugs. The first one is as designed. The second I consider another feature request. It's a bit more esoteric than the current check, but it's still a reasonable request.

Thank you.   Yes, this is a feature request.

My suggestion is to emit an error for `Y0` that says something like: "__init__ of non-final class doesn't call super`.  You can't know that the parent is `object` as you say since you don't know that some other class like `D` won't come along and introduce a different super class.

This will force either `Y0` to call super, or else the user can apply `typing.final` to `Y0`.

**Comment by erictraut** [comment]
The class hierarchy of `Y0` is well established even if it's not final, so I don't see a need to ever emit an error there. Mixing `Y0` with other classes in a multiple-inheritance situation doesn't change its hierarchy.

**Comment by NeilGirdhar** [comment]
> The class hierarchy of Y0 is well established even if it's not final, so I don't see a need to ever emit an error there

I'm not sure exactly what you mean by "the hierarchy is established".  Do you agree that Y0's super class is unknown?

**Comment by erictraut** [comment]
No, I don't agree with that. `Y0` has one superclass: `object`. That's established, and it doesn't change when `Y0` is subclassed.

Are you thinking that superclasses can somehow be dynamically inserted into the MRO? That wouldn't work. What if that newly-inserted class had an `__init__` method that required additional parameters? `Y0` would have no idea what arguments needed to be passed to make this mystery superclass happy.

It's the responsibility of a subclass to call all of the `__init__` methods of its parent classes. Only it can know what arguments need to be passed to each.

**Comment by NeilGirdhar** [comment]
> Y0 has one superclass: object. That's established, and it doesn't change when Y0 is subclassed.

Sorry, but  you're mistaken.  `Y0`'s super is `S` in `D`'s MRO.

>  What if that newly-inserted class had an __init__ method that required additional parameters? Y0 would have no idea what arguments needed to be passed to make this mystery superclass happy.

In this case, you're supposed to pass parameters as keywords to keep the superclass happy.  For example:
```python
from typing import Any

class X:
    def __init__(self, x: int, *args: Any, **kwargs: Any) -> None:  # no error!
        super().__init__(*args, **kwargs)
        self.x = x

class Y(X):
    def __init__(self, y: int, *args: Any, **kwargs: Any) -> None:  # no error!
        super().__init__(*args, **kwargs)
        self.y = y

class Z(X):
    def __init__(self, z: int, *args: Any, **kwargs: Any) -> None:
        super().__init__(*args, **kwargs)
        self.z = z

class D(Y, Z):
    pass

d = D(x=1, y=2, z=3)
print(d.x, d.y, d.z)
```
> It's the responsibility of a subclass to call all of the __init__ methods of its parent classes. Only it can know what arguments need to be passed to each.

Sorry, this is incorrect.  A class should call its superclass—not its parent classes.  In the snippet above, `D` does not call `Y.__init__` _and_ `Z.__init__`; it only calls `super().__init__`.  It doesn't even know which class `super()` refers to.  Its job is merely to forward the appropriate parameters to that super-class.  Love it or hate it, this is how cooperative multiple inheritance works.

**Comment by erictraut** [comment]
Thanks for the explanation. I hadn't fully appreciated the implications of the MRO algorithm. I see that you are correct in that it can effectively "insert" new superclasses for a class.

This is pretty messed up. I'll add this to my very long list of reasons that multiple inheritance is a terrible idea that should never have been adopted by Python or other dynamic languages.

Looking at my team's Python code base, I can now see that most of our code that involves multiple inheritance is extremely fragile, and I'm quite amazed that we haven't experienced more problems with it.

Correct me if I'm wrong, but I think this has the following implications:
1. Any class that is not marked `@final` may have superclasses that cannot be known statically if it happens to be used as a multi-inheritance base class. That makes any static type analysis involving method or variable resolution suspect.
2. Any class that has multiple base classes (regardless of whether it's marked final) should always include an `__init__` method that calls through to `super().__init__`.
3. Any class that is not final should always include an `__init__` method that calls through to `super().__init__`.
4. All `__init__` methods in a non-final class (i.e. any class that is potentially going to be used as a multi-inheritance base class) should always accept arbitrary positional and keyword arguments (`*args: Any` and `**kwargs: Any`) and pass these along as arguments to the `super().__init__` call.

Rule 4 effectively means that static type checking of arguments passed to a constructor of any non-final class will necessarily be incomplete because all such `__init__` methods accept arbitrary parameters of arbitrary types.

I'll point out that my previous misunderstanding was bolstered by various tutorials on the internet that attempted to provide advice about how to use `__init__` in a multi-inheritance scenario. These tutorials and articles are mostly wrong or at best misleading. That means even supposed "experts" do not understand the implications. The average Python developer doesn't have a prayer of getting this right.

In short, this is extremely complicated, very few Python developers understand the implications of using multiple inheritance, most Python code that does use multiple inheritance is extremely fragile by nature, and type checkers cannot do much to help here because the above rules effectively "blind" a type checker.

Maybe I should simply add a diagnostic check that reports an error any time multiple inheritance is used in a code base. ;)

**Comment by NeilGirdhar** [comment]
> This is pretty messed up. I'll add this to my very long list of reasons that multiple inheritance is a terrible idea that should never have been adopted by Python or other dynamic languages.

You're not alone in your reaction.  Personally, I've gotten accustomed to Python's coöperative multiple inheritance.  I don't find it confusing anymore, although I do agree with you that it should be avoided unless it's truly the best solution.  I think it is the best solution in some rare cases.

> Any class that is not marked @final may have superclasses that cannot be known statically if it happens to be used as a multi-inheritance base class.

Yes.

> That makes any static type analysis involving method or variable resolution suspect.

You have much more insight into static type analysis than I do. I don't know why this would be, but could you elaborate or give some examples?  If it's accessible, I'd like to understand this better.

> Any class that has multiple base classes (regardless of whether it's marked final) should always include an __init__ method that calls through to super().__init__.

Right its `__init__` should call super, _or_ it can have no `__init__` at all.  Not having an `__init__` effectively falls through to `super().__init__`.

> Any class that is not final and includes an __init__ method should call through to super().__init__.

Yes, this is my opinion too.

> An __init__ method in a non-final class (i.e. any class that is potentially going to be used as a multi-inheritance base class) should always accept arbitrary positional and keyword arguments (*args: Any and **kwargs: Any) and pass these along as arguments to the super().__init__ call.

I used to think this, but now I think it's okay for classes to forward whatever they want.  If you don't forward positional arguments, then you're not compatible with base classes that accept positional-only parameters.  If you don't forward keyword arguments, then you're not compatible with base classes that accept keyword-only parameters.  If you don't forward either one, you're not compatible with base classes that accept any required parameters.

Classes are free to be incompatible with potential future base classes.  You are right that it does limit how the classes can be used.

> type checkers cannot do much to help here because the above rules effectively "blind" a type checker.

I'm not sure what you mean here.  If you're talking about the propagation of arguments, then I hope that there will be a more advanced version of `TypeDict`.  Maybe something like:
```python
class C(D):
    def __init__(self, x: int, **kwargs: SuperKwargs) -> None:
        super().__init__(**kwargs)
        self.x = x

class F(D):
    def __init__(self, w: int, **kwargs: SuperKwargs) -> None
        super().__init__(**kwargs)
        self.w = w

class E(C, F):
    pass

c = C(x=1, y=2, z=3)  # Line 1.
e = E(x=1, y=2, z=3, w=4)  # Line 2.
```
Line 1 should verify that y and z can be passed to D.
Line 2 should verify that y, z, w can be passed to F.

What do you think?

> Maybe I should simply add a diagnostic check that reports an error any time multiple inheritance is used in a code base. ;)

Lol.

But seriously,  multiple inheritance isn't as bad as it looks on first glance.  It's just different than other languages. you have a tremendous understanding of static type checking.  I'm pretty sure you'll find a solution to this, or be able to make a typing-sig proposal if there isn't one.

**Comment by erictraut** [comment]
> You have much more insight into static type analysis than I do. I don't know why this would be, but could you elaborate or give some examples?

Many static type checks rely on the ability to statically understand the MRO for a class. If new superclasses can be inserted into that MRO by the introduction of a new class that hasn't been seen static type checker, these checks will necessarily be incomplete. Examples of such checks include:
1. Override type consistency for methods and variables
2. Dataclass field ordering
3. TypedDict field definitions
4. super() call signature compatibility
5. Verification of any behavior that depends on the presence of certain magic methods and the types they return (e.g. `__enter__`, `__iter__`, `__next__`, `__add__`, `__call__`).
6. Abstract method override validation

> Classes are free to be incompatible with potential future base classes.

Yeah, for example, almost all stdlib classes appear to not forward `*args` or `**kwargs` in their `__init__` methods, so they cannot be used flexibly in multiple inheritance hierarchies.

For now, I think I'll modify the `reportMissingSuperCall` to have the following behavior:
1. If a class is marked as `@final`, do nothing or doesn't contain an `__init__` (or other relevant magic methods), do nothing.
2. Otherwise verify that the method does one of the following: a) calls `super().<method>` using the zero, one, or two-argument form of `super` at least once, or b) calls `<BaseClass>.<method>` at least once. If none of these are found in the method body, emit a diagnostic. If at least one of them is found, don't emit a diagnostic. The assumption is that if the author has done at least one of these, they have at least thought about the problem. Their solution may still be fragile, but at least they've thought about it.

I think this generally aligns with your recommendations above.

**Comment by NeilGirdhar** [comment]
Yes, I love your solution.  Thank you, this will help me catch bugs in my code.

**Comment by erictraut** [comment]
This change will be included in the next release.

I suspect that you will be one of the few people to ever use this check. When I enable it in my team's code base (which contains about 500K lines), it generates 380 errors, all of which are probably false positives.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.210, which I just published. It will also be included in the next release of pylance.

**Comment by NeilGirdhar** [comment]
@erictraut I recently saw [this](https://docs.pytest.org/en/stable/deprecations.html#uncooperative-constructors-deprecated) and thought you might be interested.  The check you implemented could come in handy for this.
